[Ingest pipelines] Test pipeline enhancements

alisonelizabeth opened the pull request:
This PR improves the pipeline debugging workflow in the ingest node pipelines UI.

At this point, I consider this feature functionally complete. However, I think there is still room for UX improvements, which I will continue to work on with Michael this week. Feel free to leave any thoughts on the PR as well.

Other notes:
- There are some features that will be added as a separate PR, including the ability to simulate a pipeline from the pipeline table and/or details panel, and the ability to select an indexed document when testing a pipeline (https://github.com/elastic/kibana/issues/74796).
- Currently the `tag` field is added to the simulate API request as a way to map the results to each processor in the pipeline. If a user already has a `tag` field defined, it will be temporarily overridden for this call. I don't foresee any issues with this, but wanted to call it out.
- [ES bug/limitation]:  There is currently some unexpected behavior when testing a pipeline with a `pipeline` processor. The `verbose` response does not include the processor output in the response; it only returns the status. Also, if you test with a `pipeline` processor that results in an error, the simulate API times out with a 502 Bad Gateway error. Jake L. is out this week, but I will discuss this with him when he returns.

### How to test
1. Create an ingest node pipeline. For example:

```
PUT _ingest/pipeline/test
{
   "description": "_description",
    "processors": [
      {
        "set": {
          "field": "field1",
          "value": "value1"
        }
      },
      {
        "rename": {
          "field": "dont_exist",
          "target_field": "field1",
          "ignore_failure": true
        }
      },
      {
        "rename": {
          "field": "foofield",
          "target_field": "new_field",
          "on_failure": [
            {
              "set": {
                "field": "field2",
                "value": "value2"
              }
            }
          ]
        }
      },
      {
        "drop": {
          "if": "false"
        }
      },
      {
        "drop": {
          "if": "true"
        }
      }
    ]
}
```

2. Use the Ingest Node Pipelines UI and go to the "Edit" page. Click the "Add documents" link, add some sample documents, and run the pipeline. For example:

```
// The first document in this example should trigger the on_failure processor, while the second one should succeed.
 [
    {
      "_index": "index",
      "_id": "id1",
      "_source": {
        "foo": "bar"
      }
    },
   {
      "_index": "index",
      "_id": "id2",
      "_source": {
        "foo": "baz",
        "foofield": "bar"
      }
    }
  ]
```
3. Verify the results:
   - Verify the per processor status
   - View the "Output" tab per processor
   - Update/move/delete processors and verify simulation automatically runs and updates the status
   - Change the documents dropdown and verify the results

### Screenshots
<img width="1282" alt="Screen Shot 2020-08-17 at 3 45 52 PM" src="https://user-images.githubusercontent.com/5226211/90437663-c0aa5900-e0a0-11ea-94c2-a908df58383c.png">

<img width="1253" alt="Screen Shot 2020-08-17 at 3 16 26 PM" src="https://user-images.githubusercontent.com/5226211/90437624-b0927980-e0a0-11ea-9d6a-a34d9ded3a10.png">

<img width="772" alt="Screen Shot 2020-08-17 at 3 17 03 PM" src="https://user-images.githubusercontent.com/5226211/90438464-00be0b80-e0a2-11ea-8341-00d398cb0c1e.png">

### Release note
This features refines the debugging user experience when creating or editing an ingest node pipeline in the existing Ingest Node Pipelines UI. Once a sample document(s) is provided, the pipeline is executed. The UI highlights the status of each processor, and shows the user how their sample documents change shape at each step in the pipeline.

kibanamachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/67962/)
* Commit: b328e84568fbdb8824f64ba0bdf575230e5c8f73
* This comment will update when the build is complete

### Failed CI Steps
* [Check Types](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/67962/execution/node/372/log)

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/67962/","number":"67962","commit":"b328e84568fbdb8824f64ba0bdf575230e5c8f73"}]}
PIPELINE-->

kibanamachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/68034/)
* Commit: 1f2ca4a224483a48b6edd5f68462d19dbd926fd5
* This comment will update when the build is complete

### Failed CI Steps
* [Check i18n](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/68034/execution/node/388/log)

### History
* :broken_heart: [Build #67962](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/67962/) failed b328e84568fbdb8824f64ba0bdf575230e5c8f73

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/67962/","number":"67962","commit":"b328e84568fbdb8824f64ba0bdf575230e5c8f73"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/68034/","number":"68034","commit":"1f2ca4a224483a48b6edd5f68462d19dbd926fd5"}]}
PIPELINE-->

kibanamachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/68327/)
* Commit: d60d447c07e7f4ac76e8ed65d0222a7f578ac236
* This comment will update when the build is complete

### Failed CI Steps
* [Check i18n](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/68327/execution/node/388/log)

### History
* :broken_heart: [Build #68034](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/68034/) failed 1f2ca4a224483a48b6edd5f68462d19dbd926fd5
* :broken_heart: [Build #67962](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/67962/) failed b328e84568fbdb8824f64ba0bdf575230e5c8f73

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/67962/","number":"67962","commit":"b328e84568fbdb8824f64ba0bdf575230e5c8f73"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/68034/","number":"68034","commit":"1f2ca4a224483a48b6edd5f68462d19dbd926fd5"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/68327/","number":"68327","commit":"d60d447c07e7f4ac76e8ed65d0222a7f578ac236"}]}
PIPELINE-->

kibanamachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/68517/)
* Commit: d69698fcca4eee6baa6646b2605eebb3ce986ca8
* This comment will update when the build is complete

### Failed CI Steps
* [X-Pack Jest Unit Tests](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/68517/execution/node/342/log)

### History
* :broken_heart: [Build #68327](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/68327/) failed d60d447c07e7f4ac76e8ed65d0222a7f578ac236
* :broken_heart: [Build #68034](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/68034/) failed 1f2ca4a224483a48b6edd5f68462d19dbd926fd5
* :broken_heart: [Build #67962](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/67962/) failed b328e84568fbdb8824f64ba0bdf575230e5c8f73

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/67962/","number":"67962","commit":"b328e84568fbdb8824f64ba0bdf575230e5c8f73"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/68034/","number":"68034","commit":"1f2ca4a224483a48b6edd5f68462d19dbd926fd5"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/68327/","number":"68327","commit":"d60d447c07e7f4ac76e8ed65d0222a7f578ac236"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/68517/","number":"68517","commit":"d69698fcca4eee6baa6646b2605eebb3ce986ca8"}]}
PIPELINE-->

kibanamachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/68562/)
* Commit: 8144d9762fbf37b5495245ceeb576524fd4c6b1c
* This comment will update when the build is complete

### Failed CI Steps
* [Check Types](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/68562/execution/node/364/log)

### History
* :broken_heart: [Build #68517](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/68517/) failed d69698fcca4eee6baa6646b2605eebb3ce986ca8
* :broken_heart: [Build #68327](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/68327/) failed d60d447c07e7f4ac76e8ed65d0222a7f578ac236
* :broken_heart: [Build #68034](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/68034/) failed 1f2ca4a224483a48b6edd5f68462d19dbd926fd5
* :broken_heart: [Build #67962](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/67962/) failed b328e84568fbdb8824f64ba0bdf575230e5c8f73

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/67962/","number":"67962","commit":"b328e84568fbdb8824f64ba0bdf575230e5c8f73"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/68034/","number":"68034","commit":"1f2ca4a224483a48b6edd5f68462d19dbd926fd5"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/68327/","number":"68327","commit":"d60d447c07e7f4ac76e8ed65d0222a7f578ac236"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/68517/","number":"68517","commit":"d69698fcca4eee6baa6646b2605eebb3ce986ca8"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/68562/","number":"68562","commit":"8144d9762fbf37b5495245ceeb576524fd4c6b1c"}]}
PIPELINE-->

kibanamachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/68576/)
* Commit: c4c40c0c9f0b86eba146543e1c32fa6a1ed9dae1

### Build metrics
#### @kbn/optimizer bundle module count
| id | [value](https://github.com/elastic/kibana/commit/c4c40c0c9f0b86eba146543e1c32fa6a1ed9dae1) | diff | [baseline](https://github.com/elastic/kibana/commit/156692cbfbf62309d23190baad5651544ed737c7) |
| --- | --- | --- | --- |
| ingestPipelines | 404 | +14 | 390 |

#### async chunks size
| id | [value](https://github.com/elastic/kibana/commit/c4c40c0c9f0b86eba146543e1c32fa6a1ed9dae1) | diff | [baseline](https://github.com/elastic/kibana/commit/156692cbfbf62309d23190baad5651544ed737c7) |
| --- | --- | --- | --- |
| ingestPipelines | 671.3KB | +27.8KB | 643.5KB |

#### page load bundle size
| id | [value](https://github.com/elastic/kibana/commit/c4c40c0c9f0b86eba146543e1c32fa6a1ed9dae1) | diff | [baseline](https://github.com/elastic/kibana/commit/156692cbfbf62309d23190baad5651544ed737c7) |
| --- | --- | --- | --- |
| ingestPipelines | 32.7KB | -6.0B | 32.7KB |

### History
* :broken_heart: [Build #68562](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/68562/) failed 8144d9762fbf37b5495245ceeb576524fd4c6b1c
* :broken_heart: [Build #68517](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/68517/) failed d69698fcca4eee6baa6646b2605eebb3ce986ca8
* :broken_heart: [Build #68327](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/68327/) failed d60d447c07e7f4ac76e8ed65d0222a7f578ac236
* :broken_heart: [Build #68034](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/68034/) failed 1f2ca4a224483a48b6edd5f68462d19dbd926fd5
* :broken_heart: [Build #67962](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/67962/) failed b328e84568fbdb8824f64ba0bdf575230e5c8f73

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/67962/","number":"67962","commit":"b328e84568fbdb8824f64ba0bdf575230e5c8f73"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/68034/","number":"68034","commit":"1f2ca4a224483a48b6edd5f68462d19dbd926fd5"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/68327/","number":"68327","commit":"d60d447c07e7f4ac76e8ed65d0222a7f578ac236"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/68517/","number":"68517","commit":"d69698fcca4eee6baa6646b2605eebb3ce986ca8"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/68562/","number":"68562","commit":"8144d9762fbf37b5495245ceeb576524fd4c6b1c"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/68576/","number":"68576","commit":"c4c40c0c9f0b86eba146543e1c32fa6a1ed9dae1"}]}
PIPELINE-->

alisonelizabeth commented on x-pack/plugins/ingest_pipelines/public/application/components/pipeline_processors_editor/serialize.ts:
```diff
@@ -15,35 +15,46 @@ export interface SerializeResult {
   if (onFailure?.length) {
-    outProcessor[type].on_failure = convertProcessors(onFailure);
-  } else if (onFailure) {
-    outProcessor[type].on_failure = [];
```
@jloleysens can you verify I didn't introduce any regressions by deleting this line? It was causing issues when calling the simulate API, as the `on_failure` field cannot be an empty array.

jloleysens replied:
Ah ok, I _think_ this should be fine. The thinking here was that we return to the caller the same set of values that was passed in - so if they passed in an empty array, we would return one. That being said, I am happy to collapse that value because the ES API accepts undefined for that value.

Not accepting an empty array is kinda weird though!

elasticmachine commented:
Pinging @elastic/es-ui (Team:Elasticsearch UI)

kibanamachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/68628/)
* Commit: 759a000fbd828e269488bbf5bae3cc637a991bf8

### Build metrics
#### @kbn/optimizer bundle module count
| id | [value](https://github.com/elastic/kibana/commit/759a000fbd828e269488bbf5bae3cc637a991bf8) | diff | [baseline](https://github.com/elastic/kibana/commit/156692cbfbf62309d23190baad5651544ed737c7) |
| --- | --- | --- | --- |
| ingestPipelines | 404 | +14 | 390 |

#### async chunks size
| id | [value](https://github.com/elastic/kibana/commit/759a000fbd828e269488bbf5bae3cc637a991bf8) | diff | [baseline](https://github.com/elastic/kibana/commit/156692cbfbf62309d23190baad5651544ed737c7) |
| --- | --- | --- | --- |
| ingestPipelines | 672.5KB | +28.9KB | 643.5KB |

#### page load bundle size
| id | [value](https://github.com/elastic/kibana/commit/759a000fbd828e269488bbf5bae3cc637a991bf8) | diff | [baseline](https://github.com/elastic/kibana/commit/156692cbfbf62309d23190baad5651544ed737c7) |
| --- | --- | --- | --- |
| ingestPipelines | 32.7KB | -6.0B | 32.7KB |

### History
* :green_heart: [Build #68576](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/68576/) succeeded c4c40c0c9f0b86eba146543e1c32fa6a1ed9dae1
* :broken_heart: [Build #68562](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/68562/) failed 8144d9762fbf37b5495245ceeb576524fd4c6b1c
* :broken_heart: [Build #68517](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/68517/) failed d69698fcca4eee6baa6646b2605eebb3ce986ca8
* :broken_heart: [Build #68327](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/68327/) failed d60d447c07e7f4ac76e8ed65d0222a7f578ac236
* :broken_heart: [Build #68034](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/68034/) failed 1f2ca4a224483a48b6edd5f68462d19dbd926fd5

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/68034/","number":"68034","commit":"1f2ca4a224483a48b6edd5f68462d19dbd926fd5"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/68327/","number":"68327","commit":"d60d447c07e7f4ac76e8ed65d0222a7f578ac236"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/68517/","number":"68517","commit":"d69698fcca4eee6baa6646b2605eebb3ce986ca8"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/68562/","number":"68562","commit":"8144d9762fbf37b5495245ceeb576524fd4c6b1c"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/68576/","number":"68576","commit":"c4c40c0c9f0b86eba146543e1c32fa6a1ed9dae1"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/68628/","number":"68628","commit":"759a000fbd828e269488bbf5bae3cc637a991bf8"}]}
PIPELINE-->

jloleysens commented on x-pack/plugins/ingest_pipelines/public/application/components/pipeline_processors_editor/serialize.ts:
```diff
@@ -15,35 +15,46 @@ export interface SerializeResult {
 };
 
-export const serialize = ({ processors, onFailure }: SerializeArgs): SerializeResult => {
+export const serialize = (
```
nit; a default value and renaming to `copyIdToTag` could be extra clear :)

Then we document like

```ts
/**
  * param@ processor: The deserialized processor to convert
  * param@ copyIdToTag: For simulation, we add the "tag" field equal to the internal processor id so that we can map the simulate results to each processor
  */
const serialize = (processor: ProcessorInternal, copyIdToTag: boolean = false)
```

I think doing this on interface could be even better, but understand that is more of a refactor:

```ts
interface SerializeArgs {
  /**
     * The deserialized processor to convert
     */
  processor: ProcessorInternal;

  /**
     * For simulation, we add the "tag" field equal to the internal processor id so that we can map the simulate results to each processor
     */
  copyIdToTag: boolean;
}

export const serialize = ({
  pipeline: { processors, onFailure },
  copyIdToTag = false,
}: SerializeArgs): SerializeResult => {

```

alisonelizabeth commented on x-pack/plugins/ingest_pipelines/public/application/components/pipeline_processors_editor/deserialize.ts:
```diff
@@ -58,3 +58,48 @@ export const deserialize = ({ processors, onFailure }: DeserializeArgs): Deseria
+        const result = { ...currentResult };
+        const resultId = result.tag;
+
+        if (index !== 0) {
```
The first processor result would not have a previous result 😄 . I can update the code comment below to be more clear.

I also realized earlier today that this logic may need to be more complex. For example, if the previous processor was skipped there would also not be a previous result. Interested in your thoughts here - is this OK, or does it make more sense to display the last successful processor output?

alisonelizabeth commented on x-pack/plugins/ingest_pipelines/public/application/components/pipeline_form/processors_header.tsx:
```diff
@@ -4,68 +4,146 @@
+  const {
+    state: { processors },
+  } = usePipelineProcessorsContext();
+  const { testPipelineData, setCurrentTestPipelineData } = useTestPipelineContext();
```
Great point. I agree; I think this would be cleaner. I think I may have started in this direction, but can't remember now why I pivoted 🤔 . I'll look into refactoring this.

alisonelizabeth replied:
I ended up creating a component called `TestPipelineActions` that renders the add documents button/documents dropdown and the output button. The pipeline processor editor exports this as well as the `DocumentsDropdown`, since the dropdown is also reused in the output tab of each processor. Let me know what you think!

alisonelizabeth commented:
@jloleysens thanks for the very helpful review! I will address the feedback today/tomorrow. 

In the meantime, I'm not able to reproduce the error you are seeing. Are you running with the latest ES snapshot? Maybe we can zoom tomorrow if you're still having issues.

jloleysens commented:
> Are you running with the latest ES snapshot? Maybe we can zoom tomorrow if you're still having issues.

Ah that could be the case! I thought my snapshot was recent enough, but it might be the case that it is not. Will re-test :+1:

alisonelizabeth commented:
@jloleysens I believe I addressed all of your feedback. Let me know what you think, and reach out if you still run into issues testing. Thanks!

kibanamachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/69070/)
* Commit: 66a6ff3a3a2330bb7c9336155b5fc5415225755e
* This comment will update when the build is complete

### Failed CI Steps
* [X-Pack Jest Unit Tests](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/69070/execution/node/350/log)

### History
* :green_heart: [Build #68628](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/68628/) succeeded 759a000fbd828e269488bbf5bae3cc637a991bf8
* :green_heart: [Build #68576](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/68576/) succeeded c4c40c0c9f0b86eba146543e1c32fa6a1ed9dae1
* :broken_heart: [Build #68562](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/68562/) failed 8144d9762fbf37b5495245ceeb576524fd4c6b1c
* :broken_heart: [Build #68517](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/68517/) failed d69698fcca4eee6baa6646b2605eebb3ce986ca8
* :broken_heart: [Build #68327](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/68327/) failed d60d447c07e7f4ac76e8ed65d0222a7f578ac236

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/68327/","number":"68327","commit":"d60d447c07e7f4ac76e8ed65d0222a7f578ac236"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/68517/","number":"68517","commit":"d69698fcca4eee6baa6646b2605eebb3ce986ca8"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/68562/","number":"68562","commit":"8144d9762fbf37b5495245ceeb576524fd4c6b1c"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/68576/","number":"68576","commit":"c4c40c0c9f0b86eba146543e1c32fa6a1ed9dae1"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/68628/","number":"68628","commit":"759a000fbd828e269488bbf5bae3cc637a991bf8"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/69070/","number":"69070","commit":"66a6ff3a3a2330bb7c9336155b5fc5415225755e"}]}
PIPELINE-->

jloleysens reviewed (approved):
Thanks for addressing the PR feedback @alisonelizabeth ! This is looking really great.

I have a few UX related questions that I think we can address in (a) follow up PR(s):

As a user, it wasn't exactly clear to me how I could get out of live testing mode once I had added documents to test with. This is not a major issue but I thought could merit being intentional about in our design. The only instance I think something like this could be useful in is when a user wants to silence constant feedback or when they are working under high network latency conditions. I tried clearing the docs in the flyout but the form prevented me from doing so. I'd imagine clearing the docs should also clear all of the current processor statuses.

<img width="677" alt="Screenshot 2020-08-19 at 16 04 21" src="https://user-images.githubusercontent.com/8155004/90646434-7e675000-e237-11ea-9f22-0d39f1e464e5.png">

I understand that being "dropped" is a state we get back from ES simulate, but perhaps in this case it would be better suited if the drop processor had a green tick (i.e., it did it's job).

<img width="412" alt="Screenshot 2020-08-19 at 16 06 07" src="https://user-images.githubusercontent.com/8155004/90646950-12391c00-e238-11ea-9425-611de9ecb497.png">

## Bug

I don't mind if we address this in follow up PR, but if I submit a simulation with a badly configured processor currently the processor states do not update and the user does not receive an indication that their processor is incorrectly configured unless they try to create the pipeline or open the output flyout. With the following 👇🏻 configuration for "Set" and documents loaded up the UI still rendered a green tick (from my initial good configuration).

<img width="1289" alt="Screenshot 2020-08-19 at 16 26 08" src="https://user-images.githubusercontent.com/8155004/90647549-b6bb5e00-e238-11ea-9811-d120562d6459.png">

Perhaps we should render the entire pipeline's processors in "not run" status when simulate returns a 400.

## Component Integration Tests

I think it would be good to add them as the very next thing so that following PRs can benefit from them :+1:

alisonelizabeth commented:
Thanks @jloleysens for the second review!

> As a user, it wasn't exactly clear to me how I could get out of live testing mode once I had added documents to test with. This is not a major issue but I thought could merit being intentional about in our design. The only instance I think something like this could be useful in is when a user wants to silence constant feedback or when they are working under high network latency conditions. I tried clearing the docs in the flyout but the form prevented me from doing so. I'd imagine clearing the docs should also clear all of the current processor statuses.

This is a great point and not something I had thought about yet. I'll work with Michael on this and address in a separate PR.

> I understand that being "dropped" is a state we get back from ES simulate, but perhaps in this case it would be better suited if the drop processor had a green tick (i.e., it did it's job).

I think I originally only showed success/fail statuses per processor (i.e., error_ignored, skipped, and dropped == success). However, when I held the initial stakeholder review, feedback I received was that it would be helpful to call out each status. The icons I currently have are probably not the best representation. I've mentioned this to Michael already, and he's going to look into it.

> I don't mind if we address this in follow up PR, but if I submit a simulation with a badly configured processor currently the processor states do not update and the user does not receive an indication that their processor is incorrectly configured unless they try to create the pipeline or open the output flyout. With the following 👇🏻 configuration for "Set" and documents loaded up the UI still rendered a green tick (from my initial good configuration).
 
Good catch! For now, I reset the output when there is an error (back to "inactive" state), but I think there's probably more that could be done here to help the user and possibly surface the error.

> I think it would be good to add them as the very next thing so that following PRs can benefit from them 👍

Agreed! Plan to work on tests next.

kibanamachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/69152/)
* Commit: bde50929dd0783a75f7067776661e930ffabea95

### Build metrics
#### @kbn/optimizer bundle module count
| id | [value](https://github.com/elastic/kibana/commit/bde50929dd0783a75f7067776661e930ffabea95) | diff | [baseline](https://github.com/elastic/kibana/commit/32bd45bfbbea46203e7287f4ef046c7338789dc9) |
| --- | --- | --- | --- |
| ingestPipelines | 405 | +15 | 390 |

#### async chunks size
| id | [value](https://github.com/elastic/kibana/commit/bde50929dd0783a75f7067776661e930ffabea95) | diff | [baseline](https://github.com/elastic/kibana/commit/32bd45bfbbea46203e7287f4ef046c7338789dc9) |
| --- | --- | --- | --- |
| ingestPipelines | 672.8KB | +29.3KB | 643.4KB |

#### page load bundle size
| id | [value](https://github.com/elastic/kibana/commit/bde50929dd0783a75f7067776661e930ffabea95) | diff | [baseline](https://github.com/elastic/kibana/commit/32bd45bfbbea46203e7287f4ef046c7338789dc9) |
| --- | --- | --- | --- |
| ingestPipelines | 32.7KB | -6.0B | 32.7KB |

### History
* :broken_heart: [Build #69070](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/69070/) failed 66a6ff3a3a2330bb7c9336155b5fc5415225755e
* :green_heart: [Build #68628](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/68628/) succeeded 759a000fbd828e269488bbf5bae3cc637a991bf8
* :green_heart: [Build #68576](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/68576/) succeeded c4c40c0c9f0b86eba146543e1c32fa6a1ed9dae1
* :broken_heart: [Build #68562](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/68562/) failed 8144d9762fbf37b5495245ceeb576524fd4c6b1c
* :broken_heart: [Build #68517](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/68517/) failed d69698fcca4eee6baa6646b2605eebb3ce986ca8

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/68517/","number":"68517","commit":"d69698fcca4eee6baa6646b2605eebb3ce986ca8"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/68562/","number":"68562","commit":"8144d9762fbf37b5495245ceeb576524fd4c6b1c"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/68576/","number":"68576","commit":"c4c40c0c9f0b86eba146543e1c32fa6a1ed9dae1"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/68628/","number":"68628","commit":"759a000fbd828e269488bbf5bae3cc637a991bf8"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/69070/","number":"69070","commit":"66a6ff3a3a2330bb7c9336155b5fc5415225755e"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/69152/","number":"69152","commit":"bde50929dd0783a75f7067776661e930ffabea95"}]}
PIPELINE-->
